Fix ApprovalOperator with SimpleAuthManager when all_admins=True

TempestShaw opened the pull request:
Fixes #59348. Added is_allowed() method to BaseAuthManager and all
  implementations to properly delegate HITL permission checks.

  - SimpleAuthManager: Returns True when simple_auth_manager_all_admins=True
  - Other managers: Check if user is in assigned_users list
  - Updated hitl.py to use auth manager's is_allowed() method

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

TempestShaw committed:
Fix ApprovalOperator with SimpleAuthManager when all_admins=True

  Fixes #59348. Added is_allowed() method to BaseAuthManager and all
  implementations to properly delegate HITL permission checks.

  - SimpleAuthManager: Returns True when simple_auth_manager_all_admins=True
  - Other managers: Check if user is in assigned_users list
  - Updated hitl.py to use auth manager's is_allowed() method

Remove duplicate import

TempestShaw committed:
Fix ApprovalOperator with SimpleAuthManager when all_admins=True

  Fixes #59348. Added is_allowed() method to BaseAuthManager and all
  implementations to properly delegate HITL permission checks.

  - SimpleAuthManager: Returns True when simple_auth_manager_all_admins=True
  - Other managers: Check if user is in assigned_users list
  - Updated hitl.py to use auth manager's is_allowed() method

Remove duplicate import

jason810496 commented on airflow-core/src/airflow/api_fastapi/auth/managers/base_auth_manager.py:
```diff
@@ -347,6 +348,15 @@ def filter_authorized_menu_items(self, menu_items: list[MenuItem], *, user: T) -
+
+        :param user_id: the user id to check
+        :param assigned_users: list of users assigned to the task
+        """
```
Would it be better to rename the method as `is_authorized_hitl_task`? Since `is_allowed` might be too ambiguous for the whole Auth Manger context.

potiuk replied:
yeah. I think we need to also wait for @vincbeck to make some decisions on the best API to use. I honestly do not think that leaking "HITLUser" to Auth Manager is good choice. Auth Manager should maybe understand HITL as the concept, but imho it should get only the user_ids not on HitlUser objects.

potiuk replied:
Also that is a new functionality and some thinking on how we communicatee it (newsfragment) and what should be default behaviour for Auth Managers that do not implement it should be.

vincbeck replied:
Since there is a default implementation in the base auth manager, do we really need it?

jason810496 commented on providers/amazon/src/airflow/providers/amazon/aws/auth_manager/aws_auth_manager.py:
```diff
@@ -247,6 +248,14 @@ def _has_access_to_menu_item(request: IsAuthorizedRequest):
+
+        User must be in assigned_users list.
+        """
+        return any(user["id"] == user_id for user in assigned_users)
```
How about moving the `any(user["id"] == user_id for user in assigned_users)` condition to `BaseAuthManager`? So that AWS, FAB, KeyCloak manger could reuse it and only SimpleAuthManger need to override the logic.

potiuk replied:
Yeah. That would be my choice too.

TempestShaw replied:
I’ve made the requested changes. Please let me know if anything else needs adjustment.

jason810496 reviewed (commented):
Nice! Thanks for the fix and it LGTM overall.

TempestShaw committed:
Refactor HITL authorization methods across auth managers to improve consistency and remove redundant code

TempestShaw committed:
Refactor HITL authorization methods across auth managers to improve consistency and remove redundant code

jason810496 commented on airflow-core/src/airflow/api_fastapi/auth/managers/simple/simple_auth_manager.py:
```diff
@@ -283,6 +286,27 @@ def filter_authorized_menu_items(
     ) -> list[MenuItem]:
         return menu_items
 
+    def is_allowed(self, user_id: str, assigned_users: Sequence[HITLUser]) -> bool:
```
It seems this one is missing 

```suggestion
    def is_authorized_hitl_task(self, user_id: str, assigned_users: Sequence[HITLUser]) -> bool:
```

jason810496 reviewed (commented):
Thanks for the update!

> Also that is a new functionality and some thinking on how we communicatee it (newsfragment) and what should be default behaviour for Auth Managers that do not implement it should be.

It would be nice to add the `59399.feature.rst` newsfragment from `airflow-core/newsfragments/template.significant.rst` template.

Thanks!

vincbeck commented on airflow-core/src/airflow/api_fastapi/auth/managers/base_auth_manager.py:
```diff
@@ -347,6 +348,18 @@ def filter_authorized_menu_items(self, menu_items: list[MenuItem], *, user: T) -
         :param user: the user
         """
 
+    def is_authorized_hitl_task(self, user_id: str, assigned_users: Sequence[HITLUser]) -> bool:
```
Let's try to follow the other APIs pattern. Also, I am not a big fan of having `HITLUser` as dependency of the base auth manager. In fact, we only need the user ID.

```suggestion
    def is_authorized_hitl_task(self, *, assigned_users: set[str], user: T) -> bool:
```

vincbeck replied:
This method should also have unit tests

vincbeck commented on airflow-core/src/airflow/api_fastapi/auth/managers/simple/simple_auth_manager.py:
```diff
@@ -283,6 +286,27 @@ def filter_authorized_menu_items(
+            return True
+
+        # Check if user is in the assigned_users list
+        allowed = any(user["id"] == user_id for user in assigned_users)
```
We should fallback to the default implementation instead of rewriting it? `return super().is_authorized_hitl_task(...)`

TempestShaw committed:
Refactor is_authorized_hitl_task signature

- Update method to use keyword-only parameters
- Take full user object instead of just user_id
- Add unit tests for BaseAuthManager and SimpleAuthManager
- Update hitl.py call site to match new signature

TempestShaw committed:
Refactor is_authorized_hitl_task signature

- Update method to use keyword-only parameters
- Take full user object instead of just user_id
- Add unit tests for BaseAuthManager and SimpleAuthManager
- Update hitl.py call site to match new signature

Lee-W reviewed (commented):
overall looks good. But would like to confirm with @vincbeck whether it's possible to have more than one user with the same ID (I guess not?) if so, should we use `id, name` pair to check instead

pierrejeambrun reviewed (approved):
LGTM

vincbeck commented on airflow-core/docs/core-concepts/auth-manager/index.rst:
```diff
@@ -141,6 +141,7 @@ These authorization methods are:
 * ``is_authorized_asset_alias``: Return whether the user is authorized to access Airflow asset aliases. Some details about the asset alias can be provided (e.g. the asset alias ID).
 * ``is_authorized_pool``: Return whether the user is authorized to access Airflow pools. Some details about the pool can be provided (e.g. the pool name).
 * ``is_authorized_variable``: Return whether the user is authorized to access Airflow variables. Some details about the variable can be provided (e.g. the variable key).
+* ``is_authorized_hitl_task``: Return whether the user is authorized to approve or reject a Human-in-the-loop (HITL) task.
```
nit. Maybe move it at the end of the list and mention this method is optional since it has a default implementation

TempestShaw replied:
I see there is a section for `Optional methods recommended to override for optimization` at line 195. Should I move this method to that section?

vincbeck commented on airflow-core/newsfragments/59399.feature.rst:
```diff
@@ -0,0 +1 @@
+Add ``is_authorized_hitl_task()`` method to fix ApprovalOperator authorization with SimpleAuthManager
```
```suggestion
Add ``is_authorized_hitl_task()`` method to check whether a user a is authorized to approve a HITL task
```

vincbeck reviewed (commented):
2 nits but overall looks very good!

TempestShaw committed:
Update airflow-core/newsfragments/59399.feature.rst

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

TempestShaw committed:
Clarify is_authorized_hitl_task method documentation to specify custom authorization logic for HITL tasks

uranusjr commented on airflow-core/docs/core-concepts/auth-manager/index.rst:
```diff
@@ -144,7 +144,6 @@ These authorization methods are:
 * ``is_authorized_view``: Return whether the user is authorized to access a specific view in Airflow. The view is specified through ``access_view`` (e.g. ``AccessView.CLUSTER_ACTIVITY``).
 * ``is_authorized_custom_view``: Return whether the user is authorized to access a specific view not defined in Airflow. This view can be provided by the auth manager itself or a plugin defined by the user.
 * ``filter_authorized_menu_items``: Given the list of menu items in the UI, return the list of menu items the user has access to.
-
```
This line probably shouldn’t be removed?

TempestShaw committed:
modify documentation for is_authorized_hitl_task method

jason810496 reviewed (commented):
> @Lee-W Hi! I rebased fix-hitl-auth onto main to keep it up to date, which automatically triggered additional review requests.
Just wanted to check if this rebase looks correct. Thanks!

Thanks for the rebase. We still need to resolve the conflicts again. Thanks!

<img width="836" height="696" alt="Image" src="https://github.com/user-attachments/assets/d42fb497-b3ec-4af4-9b64-22bd1272dd1d" />

jason810496 reviewed (approved):
Thanks for the update! LGTM if CI pass.

I just added backport label to `3.1.6`, we can close the backport one if we only want to release in `3.2.0`.
